odoo/models.py

```diff
@@ -6668,29 +6692,71 @@ def build_like_regex(value: str, exact: bool):
     def sorted(self, key=None, reverse=False) -> Self:
         """Return the recordset ``self`` ordered by ``key``.
 
-        :param key: either a function of one argument that returns a
-            comparison key for each record, or a field name, or ``None``, in
-            which case records are ordered according the default model's order
+        :param key:
+            It can be either of:
+
+            * -a function of one argument that returns a comparison key for each record
+            * -a string representing a comma-separated list of field names with optional
+              NULLS (FIRST|LAST), and (ASC|DESC) directions
+            * -``None``, in which case records are ordered according the default model's order
         :type key: callable or str or None
         :param bool reverse: if ``True``, return the result in reverse order
 
         .. code-block:: python3
 
             # sort records by name
             records.sorted(key=lambda r: r.name)
+            # sort records by name in descending order
+            records.sorted('name DESC, id')
+            # sort records using default order
+            records.sorted()
         """
-        if key is None:
-            if any(self._ids):
-                ids = self.search([('id', 'in', self.ids)])._ids
-            else:  # Don't support new ids because search() doesn't work on new records
-                ids = self._ids
-            ids = tuple(reversed(ids)) if reverse else ids
-        else:
-            if isinstance(key, str):
-                key = itemgetter(key)
-            ids = tuple(item.id for item in sorted(self, key=key, reverse=reverse))
+        if isinstance(key, str):
+            key = self._sorted_order_fun(key)
+        elif key is None:
+            key = self._sorted_order_fun(self._order)
+        ids = tuple(item.id for item in sorted(self, key=key, reverse=reverse))
         return self.__class__(self.env, ids, self._prefetch_ids)
 
+    def _sorted_order_fun(self, order: str) -> Callable[[BaseModel], tuple]:
+        item_makers = []
+        for order_part in order.split(','):
+            order_match = regex_order.match(order_part)
+            if not order_match:
+                raise ValueError(f"Invalid order {order!r} to sort")
+            field_name = order_match['field']
+            property_name = order_match['property']
+            reverse = (order_match['direction'] or '').upper() == 'DESC'
+            nulls = (order_match['nulls'] or '').upper()
+            if not nulls:
+                nulls = 'NULLS FIRST' if reverse else 'NULLS LAST'
+
+            field = self._fields[field_name]
+            if field.type == 'many2one':
```

Comment:
The field type of `Id` is `integer`
```python
class Id(Field[IdType | typing.Literal[False]]):
    """ Special case for field 'id'. """
    type = 'integer'
    column_type = ('int4', 'int4')
```

Reply:
I mean, if you `recs.sorted('partner_id.id')` then it should compare the values of ids of the comodel while sorting instead of looking at its `_order`.

Reply:
But `_order` doesn't allow for such syntax, does it?
```python
regex_order = re.compile(r'''
    ^
    (\s*
        (?P<term>((?P<field>[a-z0-9_]+|"[a-z0-9_]+")(\.(?P<property>[a-z0-9_]+))?(:(?P<func>[a-z_]+))?))
        (\s+(?P<direction>desc|asc))?
        (\s+(?P<nulls>nulls\ first|nulls\ last))?
        \s*
        (,|$)
    )+
    (?<!,)
    $
''', re.IGNORECASE | re.VERBOSE)
```

The field name can only be `[a-z0-9_]+` without any chaining dots.

Reply:
You have property that is accepted. It's a little bit hacky, but we accept '.id' for many2one meaning to order by the 'id' of the record instead of the default order.